[WIP] Add a button for re-checking authentication status for providers

AparnaKarve opened the pull request:
Addresses <a href=https://github.com/ManageIQ/manageiq/issues/8643#issuecomment-219093632>a part</a> of https://github.com/ManageIQ/manageiq/issues/8643.

While the other major UI changes mentioned in https://github.com/ManageIQ/manageiq/issues/8643 for the ems editors are still underway, we can move forward with this particular change which basically adds a `Re-check Authentication` button in the EMS Summary screen to update the Authentication status.

<img width="1209" alt="screen shot 2016-06-01 at 7 06 53 am" src="https://cloud.githubusercontent.com/assets/1538216/15712539/809ce17a-27c7-11e6-998e-170513680f31.png">

AparnaKarve commented:
@Fryguy @jrafanie Please review the new method `authentication_check_queue` in authentication_mixin https://github.com/ManageIQ/manageiq/pull/8912/commits/7486a143a8fbfdbf35637214c17eea0a49d968df. It's pretty much based on one of the existing `_queue` methods in that mixin.

AparnaKarve commented:
/cc @Ladas @juliancheal

AparnaKarve commented:
@miq-bot add_label wip,ui,providers

jrafanie commented on app/models/mixins/authentication_mixin.rb:
```diff
@@ -273,6 +273,27 @@ def authentication_check(*args)
     return status == :valid, details
   end
 
+  def authentication_check_queue(*args)
```
Can you use the nearly identical `authentication_check_types_queue`?

It's what we call from the schedule worker on a schedule for each model that includes the authentication mixin.

https://github.com/ManageIQ/manageiq/blob/fa42171d9cac349eea580a96c134384de949a5d2/app/models/mixins/authentication_mixin.rb#L10-L13

jrafanie replied:
If there's something you can't do, please provide a test or script demonstrating what's missing and I'll try to fix the code to do what you want.

AparnaKarve replied:
@jrafanie `authentication_check_types_queue` works just as great! Thanks for pointing me to this method.
However, the only thing that needs to happen before we can use this method is to make sure that ems types that have authentication types other than `default` should have this method available - `authentications_to_validate`.

jrafanie replied:
Good point @AparnaKarve.

The concept was that classes could dictate what authentication types it used to authenticate via `authentications_to_validate`.

Another way is to pass the types, see [here](https://github.com/ManageIQ/manageiq/blob/4bdabf303ac33dbc95466c49a7a66a081494561c/spec/models/mixins/authentication_mixin_spec.rb#L329).  If you don't specify the types, then use `authentications_to_validate` will be used.

jrafanie commented on db/fixtures/miq_product_features.yml:
```diff
@@ -364,6 +364,10 @@
+    - :name: Re-check Authentication Status 
+      :description: Re-check Authentication Status of Cloud Providers
+      :feature_type: control
+      :identifier: ems_cloud_recheck_auth_status 
```
Is this only for ems_cloud types?  I'd think it's for each provider/ems class.  This feature should be more generic or we need one for Ems, Providers, etc.

AparnaKarve replied:
@jrafanie This feature is for all ems types. I was just taking a "test drive" with `ems_cloud`...so once we are all good with the changes for `ems_cloud`, I would be adding this feature for the other ems types.

I agree that the feature is generic, however in the .yml we would have to repeat this feature individually in each ems section... so that would be `ems_infra_recheck_auth_status`, `ems_container_recheck_auth_status` etc

jrafanie replied:
👍

jrafanie commented on app/controllers/ems_common.rb:
```diff
@@ -437,6 +437,28 @@ def button
+                when "default" then _("Default")
+                when "metrics" then _("C & U Database")
+                when "amqp"    then _("AMQP")
+                else;               _("<Unknown>")
```
This case statement is already in 3 other places.  For example: https://github.com/ManageIQ/manageiq/blob/fa42171d9cac349eea580a96c134384de949a5d2/app/helpers/ems_infra_helper/textual_summary.rb#L152-L158

```
app/helpers/ems_cloud_helper/textual_summary.rb:        when "metrics" then _("C & U Database")
app/helpers/ems_infra_helper/textual_summary.rb:        when "metrics" then _("C & U Database")
app/helpers/ems_network_helper/textual_summary.rb:        when "metrics" then "C & U Database"
```

Maybe we can ask an authentication decorator for this friendly name.  cc @martinpovolny

AparnaKarve replied:
Since I plan to use `authentication_check_types_queue`, I would not be rendering the flash message for each individual authentication type now...meaning the above code can be cleaned out and only a single flash message would be displayed -
```
Re-checking Authentication status initiated for the selected Cloud Provider
```

jrafanie replied:
We should have the right abstraction here though.  But, when we're duplicating logic like this... we should ask if this is concept that needs an abstraction.

AparnaKarve commented:
@Ladas Can you review this?
I've added the `'Re-check Authentication'` button for `ems_cloud`, `ems_infra` and `ems_container`.
As discussed, also updated the alert message - https://github.com/ManageIQ/manageiq/pull/8912/commits/dd94f8b03b02615291a244bb195cf3e5d3165127

Ladas commented:
@AparnaKarve hm one question, does this recheck authentication of all attached endpoints? E.g. for OpenStack you can have default and events endpoint there, while only amqp can be rechecked

If ^ is working, it looks good to me, though I don't have a space to actually try this now.

AparnaKarve commented:
@Ladas It rechecks authentications of all attached endpoints.
Basically, whatever authentications you see on the Summary page, those will be rechecked.

For e.g - There could be a situation where an OpenStack provider has both `default` and `amqp` endpoints defined, so for that provider, both authentications will be rechecked.
Likewise, there could be an OpenStack provider with only the `default` endpoint defined, in which case only the `default` authentication will be rechecked.

So just to clarify the above, you cannot "cherry-pick" (for the lack of a better word) what authentication type to recheck.

Does that sound OK to you?

Ladas commented:
@AparnaKarve yeah, that sounds good. In that case looks :+1: to me :-)

AparnaKarve commented:
@Ladas Thanks for the confirmation above.
Now if only I could get a volunteer to test this PR :) -- I would think we *do* this for darga.
@Ladas, tbh, you are the perfect candidate to test this... just saying 😉

AparnaKarve commented:
@miq-bot remove_label wip

martinpovolny commented on app/helpers/application_helper/button/ems_infra_recheck_auth_status.rb:
```diff
@@ -0,0 +1,5 @@
+  def skip?
+    !@record.is_available?(:authentication_status)
+  end
+end
```
You don't need to create several classes with the same code. 

If the behavior you need here just checks for auth status on the record, you can just create 1 class for that and share the class for all the buttons

martinpovolny commented on app/helpers/application_helper/toolbar/ems_container_center.rb:
```diff
@@ -77,4 +77,20 @@ class ApplicationHelper::Toolbar::EmsContainerCenter < ApplicationHelper::Toolba
+      ]
+    ),
+  ])
 end
```
So the `button_group` is just a function call. There's no need to repeat the same function call in several toolbars. You can put it in the module and share the module.

abonas commented:
@AparnaKarve middleware provider seems out of scope for this feature and PR - any reason why?

AparnaKarve commented:
@abonas The main reason would be this feature was targeted for Darga.
I can create a separate PR for middleware for this feature.

abonas commented:
> @abonas The main reason would be this feature was targeted for Darga.
> I can create a separate PR for middleware for this feature.

thanks @AparnaKarve , it can definitely be a separate PR.
